### PR TITLE
Integrate strimzi-kafka-oauth 0.8.0

### PR DESCRIPTION
**Pull request by mstruk**
Signed-off-by: Marko Strukelj <marko.strukelj@gmail.com>

### Type of change

- Enhancement / new feature

### Description

This PR integrates `strimzi-kafka-oauth` 0.8.0 features.

It  deprecates the `enableECDSA` option of Kafka CR `authentication` of type `oauth`. That option used to install the BouncyCastle JCE crypto provider and thereby enable the ECDSA support. That is no longer needed, and ECDSA is supported automatically. The `keycloak-*` helper libraries and the BouncyCastle libraries are no longer bundled in `strimzi/kafka` images.

Additionally, two options have been added:
- `clientScope`
- `clientAudience`

These options allow passing a custom `scope` and `audience` to the token endpoint of the authorization server when using inter-broker authentication or for `OAuth over PLAIN` with `clientId` + `secret`.

Also, `audience` option has been added to KafkaConnect CR and KafkaBridge CR `authentication` section, where `scope` has already been available before, but not used. Now both `scope` and `audience` are propagated to KafkaBridge and KafkaConnect pods.

This PR dependes on:

strimzi/strimzi-kafka-oauth#111
strimzi/strimzi-kafka-oauth#112
strimzi/strimzi-kafka-oauth#113
strimzi/strimzi-kafka-oauth#114
strimzi/strimzi-kafka-oauth#116

### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [ ] Write tests
- [ ] Make sure all tests pass
- [ ] Update documentation
- [ ] Check RBAC rights for Kubernetes / OpenShift roles
- [ ] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally
- [ ] Reference relevant issue(s) and close them after merging
- [ ] Update CHANGELOG.md
- [ ] Supply screenshots for visual changes, such as Grafana dashboards

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -8,6 +8,7 @@
 * Added support for configuring cluster-operator's worker thread pool size that is used for various sync and async tasks
 * Add Kafka Quotas plugin with produce, consume, and storage quotas
 * Support pausing reconciliation of KafkaTopic CR with annotation `strimzi.io/pause-reconciliation`
+* The option `enableECDSA` of Kafka CR `authentication` of type `oauth` has been deprecated and is ignored. ECDSA token signature support is now always enabled without the need for Strimzi Cluster Operator installing the BouncyCastle JCE crypto provider. BouncyCastle library is no longer packaged with Strimzi Kafka images.
```
Comment (scholzj) [inline CHANGELOG.md:11]: Can you move that to the _Changes, deprecations and removals_ section below maybe?

**Comment by scholzj** [comment]
BTW: The snapshots will be here: https://oss.sonatype.org/content/repositories/snapshots ... but we need to first merge the OAuth PR

File: api/src/main/java/io/strimzi/api/kafka/model/listener/KafkaListenerAuthenticationOAuth.java [hunk 2]
```diff
@@ -53,12 +55,14 @@
...
     private boolean disableTlsHostnameVerification = false;
-    private boolean enableECDSA = false;
+    private boolean enableECDSA = true;
     private Integer maxSecondsWithoutReauthentication;
     private boolean enablePlain = false;
     private String tokenEndpointUri;
     private boolean enableOauthBearer = true;
     private String customClaimCheck;
+    private String useScope = null;
+    private String useAudience = null;
```
Comment (scholzj) [inline api/src/main/java/io/strimzi/api/kafka/model/listener/KafkaListenerAuthenticationOAuth.java:65]: `useScope` and `useAudience` sound to me like boolean values. Would it be better to call these just `audience` and `scope`? WDYT @tombentley?

File: cluster-operator/src/test/java/io/strimzi/operator/cluster/model/KafkaBrokerConfigurationBuilderTest.java [hunk 3]
```diff
@@ -1147,9 +1147,9 @@ public void testOauthConfiguration()  {
                 "ssl.endpoint.identification.algorithm=HTTPS",
                 "principal.builder.class=io.strimzi.kafka.oauth.server.OAuthKafkaPrincipalBuilder",
                 "listener.name.plain-9092.oauthbearer.sasl.server.callback.handler.class=io.strimzi.kafka.oauth.server.JaasServerOauthValidatorCallbackHandler",
-                "listener.name.plain-9092.oauthbearer.sasl.jaas.config=org.apache.kafka.common.security.oauthbearer.OAuthBearerLoginModule required unsecuredLoginStringClaim_sub=\"thePrincipalName\" oauth.valid.issuer.uri=\"http://valid-issuer\" oauth.jwks.endpoint.uri=\"http://jwks\" oauth.jwks.refresh.min.pause.seconds=\"5\" oauth.crypto.provider.bouncycastle=\"true\" oauth.username.claim=\"preferred_username\";",
+                "listener.name.plain-9092.oauthbearer.sasl.jaas.config=org.apache.kafka.common.security.oauthbearer.OAuthBearerLoginModule required unsecuredLoginStringClaim_sub=\"thePrincipalName\" oauth.valid.issuer.uri=\"http://valid-issuer\" oauth.jwks.endpoint.uri=\"http://jwks\" oauth.jwks.refresh.min.pause.seconds=\"5\" oauth.username.claim=\"preferred_username\";",
```
Comment (scholzj) [inline cluster-operator/src/test/java/io/strimzi/operator/cluster/model/KafkaBrokerConfigurationBuilderTest.java:1150]: Should we also add the scope and audience into these tests somewhere?

File: api/src/main/java/io/strimzi/api/kafka/model/listener/KafkaListenerAuthenticationOAuth.java (hunk 2 above)
Comment (mstruk) [reply api/src/main/java/io/strimzi/api/kafka/model/listener/KafkaListenerAuthenticationOAuth.java:65]: The idea here is that these are client side options used when the server acts as a client. So maybe the better name would thus be: `clientScope` and `clientAudience`. But that can again be interpreted to mean scope sent by client or audience sent by client (and contained in the token) to the broker. Then we could maybe call it `sendAudience` and `sendScope` which makes it clearer maybe that it's part of broker including these into the request. Or maybe call it `includeScope` and `includeAudience`.

File: cluster-operator/src/test/java/io/strimzi/operator/cluster/model/KafkaBrokerConfigurationBuilderTest.java (hunk 3 above)
Comment (mstruk) [reply cluster-operator/src/test/java/io/strimzi/operator/cluster/model/KafkaBrokerConfigurationBuilderTest.java:1150]: Ah yes, to make sure they are passed to Kafka Broker strimzi.properties file.

File: api/src/main/java/io/strimzi/api/kafka/model/listener/KafkaListenerAuthenticationOAuth.java (hunk 2 above)
Comment (mstruk) [reply api/src/main/java/io/strimzi/api/kafka/model/listener/KafkaListenerAuthenticationOAuth.java]: The only suggested alternative that does not suggest a boolean value here is `clientScope` and `clientAudience` so I renamed to that.

**Comment by mstruk** [comment]
Actually we also have to add `audience` to `KafkaClientAuthenticationOAuth`

File: documentation/modules/appendix_crds.adoc [hunk 4]
```diff
@@ -187,15 +187,19 @@ It must have the value `oauth` for the type `KafkaListenerAuthenticationOAuth`.
...
+|string
 |clientSecret                       1.2+<.<a|Link to Kubernetes Secret containing the OAuth client secret which the Kafka broker can use to authenticate against the authorization server and use the introspect endpoint URI.
 |xref:type-GenericSecretSource-{context}[`GenericSecretSource`]
 |customClaimCheck                   1.2+<.<a|JsonPath filter query to be applied to the JWT token or to the response of the introspection endpoint for additional token validation. Not set by default.
 |string
 |disableTlsHostnameVerification     1.2+<.<a|Enable or disable TLS hostname verification. Default value is `false`.
 |boolean
-|enableECDSA                        1.2+<.<a|Enable or disable ECDSA support by installing BouncyCastle crypto provider. Default value is `false`.
+|enableECDSA                        1.2+<.<a|*The `enableECDSA` property has been deprecated.* Enable or disable ECDSA support by installing BouncyCastle crypto provider. Since version 0.24.0 BouncyCastle crypto provider is no longer needed for ECDSA. As a result ECDSA is always enabled and BouncyCastle is not packaged with Strimzi. Value is ignored.
```
Comment (laidan6000) [inline documentation/modules/appendix_crds.adoc:202]: I would rather we didn't include any version-specific information here.

The bouncy castle libraries are no longer packaged with Strimzi, so I think you can remove the old description. How about something like this:

> *The `enableECDSA` property has been deprecated.* ECDSA support is always enabled. The BouncyCastle libraries are no longer packaged with Strimzi. Value is ignored.

**Review by laidan6000 — CHANGES_REQUESTED** [review]
Reviewed the ECDSA doc changes.

File: documentation/modules/appendix_crds.adoc (hunk 4 above)
Comment (mstruk) [reply documentation/modules/appendix_crds.adoc:202]: Ok, will change it.

File: api/src/main/java/io/strimzi/api/kafka/model/listener/KafkaListenerAuthenticationOAuth.java [hunk 5]
```diff
@@ -132,6 +136,26 @@ public void setCustomClaimCheck(String customClaimCheck) {
         this.customClaimCheck = customClaimCheck;
     }
 
+    @Description("Scope to use when making a request to authorization server's token endpoint. Used for inter-broker authentication, and for OAuth over PLAIN with `clientId` and `secret`.")
```
Comment (laidan6000) [inline api/src/main/java/io/strimzi/api/kafka/model/listener/KafkaListenerAuthenticationOAuth.java:139]:
```suggestion
    @Description("The scope to use when making requests to the authorization server's token endpoint. Used for inter-broker authentication and for configuring OAuth 2.0 over PLAIN using the `clientId` and `secret` method.")
```

File: api/src/main/java/io/strimzi/api/kafka/model/listener/KafkaListenerAuthenticationOAuth.java [hunk 6]
```diff
@@ -132,6 +136,26 @@ public void setCustomClaimCheck(String customClaimCheck) {
...
+    @JsonInclude(JsonInclude.Include.NON_NULL)
+    public String getClientScope() {
+        return clientScope;
+    }
+
+    public void setClientScope(String scope) {
+        this.clientScope = scope;
+    }
+
+    @Description("Audience to use when making a request to authorization server's token endpoint. Used for inter-broker authentication, and for OAuth over PLAIN with `clientId` and `secret`.")
```
Comment (laidan6000) [inline api/src/main/java/io/strimzi/api/kafka/model/listener/KafkaListenerAuthenticationOAuth.java:149]:
```suggestion
    @Description("The audience to use when making requests to the authorization server's token endpoint. Used for inter-broker authentication and for configuring OAuth 2.0 over PLAIN using the `clientId` and `secret` method.")
```

**Review by laidan6000 — CHANGES_REQUESTED** [review]
Made a few more suggested changes.

File: helm-charts/helm3/strimzi-kafka-operator/crds/040-Crd-kafka.yaml [hunk 7]
```diff
@@ -128,7 +128,8 @@ spec:
                                 description: Enable or disable TLS hostname verification. Default value is `false`.
                               enableECDSA:
                                 type: boolean
-                                description: Enable or disable ECDSA support by installing BouncyCastle crypto provider. Default value is `false`.
+                                description: Enable or disable ECDSA support by installing BouncyCastle crypto provider. ECDSA support is always enabled.
```
Comment (scholzj) [inline helm-charts/helm3/strimzi-kafka-operator/crds/040-Crd-kafka.yaml:131]: This file should not be changed - this is the CRD from 0.23 (all changes should be done to `packaging/...`). Can you please revert it?

File: install/cluster-operator/040-Crd-kafka.yaml [hunk 8]
```diff
@@ -174,7 +174,9 @@ spec:
                             enableECDSA:
                               type: boolean
                               description: Enable or disable ECDSA support by installing
-                                BouncyCastle crypto provider. Default value is `false`.
+                                BouncyCastle crypto provider. ECDSA support is always enabled.
```
Comment (scholzj) [inline install/cluster-operator/040-Crd-kafka.yaml:177]: This file should not be changed - this is the CRD from 0.23 (all changes should be done to `packaging/...`). Can you please revert it?

**Review by scholzj — COMMENTED** [review]
Two nits about wrongly changed CRD files - LGTM otherwise.

File: helm-charts/helm3/strimzi-kafka-operator/crds/040-Crd-kafka.yaml (hunk 7 above)
Comment (mstruk) [reply helm-charts/helm3/strimzi-kafka-operator/crds/040-Crd-kafka.yaml]: Hard to keep check on these rules ... I reverted.

File: install/cluster-operator/040-Crd-kafka.yaml (hunk 8 above)
Comment (mstruk) [reply install/cluster-operator/040-Crd-kafka.yaml]: Reverted.

**Review by scholzj — APPROVED** [review]

File: documentation/modules/oauth/proc-oauth-kafka-config.adoc [hunk 9]
```diff
@@ -96,13 +96,17 @@ spec:
...
+    audience: kafka <5>
 ----
 <1> (Optional) Disable TLS hostname verification. Default is `false`.
 <2> If the authorization server does not return a `typ` (type) claim inside the JWT token, you can apply `checkAccessTokenType: false` to skip the token type check. Default is `true`.
 <3> If you are using opaque tokens, you can apply `accessTokenIsJwt: false` so that access tokens are not treated as JWT tokens.
 <4> (Optional) The `scope` for requesting the token from the token endpoint.
 An authorization server may require a client to specify the scope.
 In this case it is `any`.
+<5> (Optional) The `audience` for requesting the token from the token endpoint.
+An authorization server may require a client to specify the scope.
```
Comment (PaulRMellor) [inline documentation/modules/oauth/proc-oauth-kafka-config.adoc:108]:
```suggestion
An authorization server may require a client to specify the audience.
```

File: documentation/modules/oauth/proc-oauth-authentication-broker-config.adoc [hunk 10]
```diff
@@ -137,6 +137,8 @@ Depending on how you apply OAuth 2.0 authentication, and the type of authorizati
...
 <9> This has to be set to support the client credentials authentication when `enablePlain` is set to true, as described in previous point.
 <10> Additional custom rules can be imposed on the JWT access token during validation by setting this to a JsonPath filter query. If the access token does not contain the necessary data, it is rejected. When using the `introspectionEndpointUri`, the custom check is applied to the introspection endpoint response JSON.
+<11> An `audience` parameter passed to the token endpoint when obtaining an access token for inter-broker authentiction or in the name of the client during `OAuth over PLAIN` with `clientId` and `secret`. This only affects the ability to obtain the token, and the content of the token, depending on the authorization server. It does not affect token validation rules by the listener.
```
Comment (PaulRMellor) [inline documentation/modules/oauth/proc-oauth-authentication-broker-config.adoc:140]:
```suggestion
<11> An `audience` parameter passed to the token endpoint when obtaining an access token for inter-broker authentiction or in the name of the client for OAuth 2.0 over PLAIN authentication with a `clientId` and `secret`. This only affects the ability to obtain the token, and the content of the token, depending on the authorization server. It does not affect token validation rules by the listener.
```

File: documentation/modules/oauth/proc-oauth-authentication-broker-config.adoc [hunk 11]
```diff
@@ -137,6 +137,8 @@ Depending on how you apply OAuth 2.0 authentication, and the type of authorizati
...
+<12> A `scope` parameter passed to the token endpoint when obtaining an access token for inter-broker authentication or in the name of the client during `OAuth over PLAIN` with `clientId` and `secret`. This only affects the ability to obtain the token, and the content of the token, depending on the authorization server. It does not affect token validation rules by the listener.
```
Comment (PaulRMellor) [inline documentation/modules/oauth/proc-oauth-authentication-broker-config.adoc:141]:
```suggestion
<12> A `scope` parameter passed to the token endpoint. A _scope_ is used when obtaining an access token for inter-broker authentication. It is also used in the name of a client for OAuth 2.0 over PLAIN client authentication using a `clientId` and `secret`. This only affects the ability to obtain the token, and the content of the token, depending on the authorization server. It does not affect token validation rules by the listener.
```

**Review by tombentley — APPROVED** [review]

File: documentation/modules/oauth/proc-oauth-authentication-broker-config.adoc [hunk 12]
```diff
@@ -126,6 +124,8 @@ Depending on how you apply OAuth 2.0 authentication, and the type of authorizati
     enablePlain: true <8>
     tokenEndpointUri: __https://OAUTH-SERVER-ADDRESS/auth/realms/external/protocol/openid-connect/token__ <9>
     customClaimCheck: "@.custom == 'custom-value'" <10>
+    clientAudience: __audience__ <11>
+    clientScope: __scope__ <12>
```
Comment (PaulRMellor) [inline documentation/modules/oauth/proc-oauth-authentication-broker-config.adoc:128]:
```suggestion
    clientAudience: __AUDIENCE__ <11>
    clientScope: __SCOPE__ <12>
```

**Review by PaulRMellor — APPROVED** [review]

**Pull request merged**